File: android/app/src/main/java/com/dibay/app/DibayCanonicalDeviceIdStore.java
package com.dibay.app;

import android.content.Context;
import android.content.SharedPreferences;

/**
 * Canonical call/push device identity — must match {@code user_devices.device_id}
 * ({@code dibay:client_instance_id} from Web register), not ANDROID_ID.
 *
 * @see docs/dibay-call-multi-device-policy.md
 */
public final class DibayCanonicalDeviceIdStore {
  private static final String PREFS = "dibay_device_identity";
  private static final String KEY = "canonical_device_id";

  private DibayCanonicalDeviceIdStore() {}

  public static void save(Context context, String deviceId) {
    if (context == null) return;
    String id = deviceId != null ? deviceId.trim() : "";
    if (id.isEmpty()) return;
    SharedPreferences prefs =
        context.getApplicationContext().getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    prefs.edit().putString(KEY, id).apply();
  }

  /** Prefer registered canonical id; empty if never registered on this install. */
  public static String resolveOrEmpty(Context context) {
    if (context == null) return "";
    SharedPreferences prefs =
        context.getApplicationContext().getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    String id = prefs.getString(KEY, "");
    return id != null ? id.trim() : "";
  }
}

File: ios/App/App/Call/DibayCanonicalDeviceIdStore.swift
import Foundation
import WebKit
import Capacitor

/**
 * Canonical call/push device identity — must match `user_devices.device_id`
 * (`dibay:client_instance_id` from Web register), not IDFV.
 *
 * @see docs/dibay-call-multi-device-policy.md
 */
enum DibayCanonicalDeviceIdStore {
  private static let defaultsKey = "dibay_canonical_device_id"
  private static let localStorageKey = "dibay:client_instance_id"

  static func save(_ deviceId: String) {
    let id = deviceId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !id.isEmpty else { return }
    UserDefaults.standard.set(id, forKey: defaultsKey)
  }

  static func resolveCached() -> String? {
    let id = (UserDefaults.standard.string(forKey: defaultsKey) ?? "")
      .trimmingCharacters(in: .whitespacesAndNewlines)
    return id.isEmpty ? nil : id
  }

  /**
   * Resolve for Native accept / answered_elsewhere.
   * Prefer UserDefaults (persisted after register); else read WebView localStorage once.
   */
  static func resolve(completion: @escaping (String?) -> Void) {
    if let cached = resolveCached() {
      completion(cached)
      return
    }
    DispatchQueue.main.async {
      guard let webView = resolveWebView() else {
        completion(nil)
        return
      }
      let js = "localStorage.getItem('\(localStorageKey)')"
      webView.evaluateJavaScript(js) { result, _ in
        let raw = (result as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !raw.isEmpty {
          save(raw)
          completion(raw)
        } else {
          completion(nil)
        }
      }
    }
  }

  private static func resolveWebView() -> WKWebView? {
    guard
      let window = UIApplication.shared.connectedScenes
        .compactMap({ $0 as? UIWindowScene })
        .flatMap({ $0.windows })
        .first(where: { $0.isKeyWindow }),
      let root = window.rootViewController as? CAPBridgeViewController,
      let webView = root.webView
    else {
      return nil
    }
    return webView
  }
}
